providers/writing_a_new_provider.md

```diff
@@ -0,0 +1,555 @@
+
+### Add your provider's SDK to the gemspec
+
+Let's add this to our provider's gemspec:
+```bash
+$ git diff
+diff --git a/manageiq-providers-awesome_cloud.gemspec b/manageiq-providers-awesome_cloud.gemspec
+index 6c228c2..91e8e71 100644
+--- a/manageiq-providers-awesome_cloud.gemspec
++++ b/manageiq-providers-awesome_cloud.gemspec
+@@ -18,6 +18,8 @@ Gem::Specification.new do |spec|
+   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
+   spec.require_paths = ["lib"]
+
++  spec.add_dependency "awesome_cloud", "~> 0.1"
++
+   spec.add_development_dependency "manageiq-style"
+   spec.add_development_dependency "simplecov"
+```
+
+Then bundle update to pull in the change
+```
+$ bundle update
+```
+
+Now that we have the gem installed we can start to write our connection code.  ManageIQ providers have to expose a `#connect` and a `#verify_credentials` method on the class and the instance.  The class method is used when adding a provider (when there is no instance record yet) and the instance methods are used after.
+
+Let's assume that Awesome Cloud requires a region, access_key, and secret_key in order to connect.
+
+```ruby
+>> ems = ManageIQ::Providers::AwesomeCloud::CloudManager.first
+>> ems.update!(:provider_region => "us-east-1")
+>> ems.update_authentication(:default => {:userid => "MY-ACCESS-KEY", :password => "MY-SECRET-KEY"})
+```
+
+Now let's add connect methods to our provider.
+
+### Connecting and verifying credentials
+
+```ruby
+class ManageIQ::Providers::AwesomeCloud::CloudManager < ManageIQ::Providers::CloudManager
+  def self.raw_connect(region, access_key, secret_key, service = "Compute")
+    require "awesome_cloud/#{service.to_s.downcase}"
+    credentials = AwesomeCloud::Credentials.new(:access_key => access_key, :secret_key => secret_key)
+
+    AwesomeCloud.const_get(service).new(:credentials => credentials, :region => region)
+  end
+
+  def self.verify_credentials(args)
+    # NOTE this args hash has a very specific format that we'll get to next
+    region = args["provider_region"]
+    default_endpoint = args.dig("authentications", "default")
+    access_key, secret_key = default_endpoint&.values_at("userid", "password")
+
+    validate_connection(raw_connect(region, access_key, secret_key))
+  end
+
+  # NOTE: You want to use the same method for validating credentials at the
+  # class-level and instance-level.  This prevents any potential issues were
+  # adding a new provider (class-level) might pass verification but after adding
+  # it (instance-lavel) it fails.
+  def self.validate_connection(connection)
+    # Perform a simple and fast operation that verifies the credentials are correct
+    !!connection.list_regions
```

Comment:
Is this the method that _must_ return a boolean?  If so, should we state that here?  I wonder there's a nice way  to ensure that at the base level

Reply:
No this can return anything, we handle the "must return a boolean" in core.